Chapter 8/Lesson 4/homework/homework.js

```
@@ -0,0 +1,107 @@
+    }
+}
+
+function isString(value) {
+    let str = "must be a string."
+    if (typeof value !== "string") {
+        return str;
+    }
+}
+
+function hasLengthBetween(a, b) {
+    let str = "must be between 1 and 100 characters long."
+    return function (value) {
+        if (value.length < a || value.length > b) {
+            return str;
+        }
+    }
+}
+
+function isNumber(value) {
+    let str = "must be a number."
+    if (typeof value !== "number") {
+        return str;
+    }
+}
+
+function isGreaterThan(number) {
+    let str = "must be greater than 0."
+    return function (value) {
+        if (value < number) {
+            return str;
+        }
+    }
+}
+
+function composeValidators(...arr) {
+    return function (fieldName) {
+        return function (value) {
+            let result;
+            for (let i = 0; i < arr.length; i++) {
+                result = arr[i](value);
+                if (result) {
+                    let message = `${fieldName} ${result}`;
+                    return message;
+                }
+            }
+        }
+    }
+}
+
+let validateName = composeValidators(
+    isRequired,
+    isString,
+    hasLengthBetween(2, 100)
+)("Name");
+
+
+let validateAge = composeValidators(
+    isNumber,
+    isGreaterThan(0)
+)("Age");
+
+
+function withValidation(arrWithFunctions) {
+    return function (func) {
+        return function () {
+            let arrWithParameters = [].concat(...arguments);
+            if (arrWithFunctions.length != arrWithParameters.length) {
+                throw new Error("The number of passed functions and arguments is different.");
+            }
+            outer: for (let i = 0; i < arrWithFunctions.length; i++) {
+                for (let j = 0; j < arrWithParameters.length; j++) {
+                    let result = arrWithFunctions[i](arrWithParameters[j]);
+                    try {
+                        if (i == j && result) throw new Error(`${result}`);
```

You don't need to catch the error, you need to throw an error outside. External code needs to handle the error.